plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,881 @@
+						  rcd->addr,
+						  rcd->buf->data, rcd->buf->len,
+						  FU_CXAUDIO_OPERATION_FLAG_VERIFY,
+						  error)) {
+			g_prefix_error (error, "failed to write @0x%04x len:0x%02x: ",
+					rcd->addr, rcd->buf->len);
+			return FALSE;
+		}
+		fu_device_set_progress_full (device, (gsize) i, (gsize) records->len);
+	}
+
+	/* in case of a full FW upgrade invalidate the old FW patch (if any)
+	 * as it may have not been done by the S37 file */
+	if (file_kind == FU_CXAUDIO_FILE_KIND_CX2070X_FW) {
+		FuCxaudioEepromPatchInfo pinfo = { 0 };
+		if (!fu_cxaudio_device_operation (self,
+						  FU_CXAUDIO_OPERATION_READ,
+						  FU_CXAUDIO_MEM_KIND_EEPROM,
+						  FU_CXAUDIO_EEPROM_PATCH_INFO_OFFSET,
+						  (guint8 *) &pinfo, sizeof(pinfo),
+						  FU_CXAUDIO_OPERATION_FLAG_NONE,
+						  error)) {
+			g_prefix_error (error, "failed to read EEPROM patch info: ");
+			return FALSE;
+		}
+		if (pinfo.PatchSignature == FU_CXAUDIO_SIGNATURE_PATCH_BYTE) {
+			memset (&pinfo, 0x0, sizeof(pinfo));
+			if (!fu_cxaudio_device_operation (self,
+							  FU_CXAUDIO_OPERATION_WRITE,
+							  FU_CXAUDIO_MEM_KIND_EEPROM,
+							  FU_CXAUDIO_EEPROM_PATCH_INFO_OFFSET,
+							  (guint8 *) &pinfo, sizeof(pinfo),
+							  FU_CXAUDIO_OPERATION_FLAG_NONE,
+							  error)) {
+				g_prefix_error (error, "failed to write empty EEPROM patch info");
+				return FALSE;
+			}
+			g_debug ("invalidated old FW patch for CX2070x (RAM) device");
+		}
+	}
+
+	/* unpark the FW */
+	if (!fu_cxaudio_device_register_clear_bit (self, FU_CXAUDIO_REG_FIRMWARE_PARK_ADDR, 7, error))
+		return FALSE;
+
+	/* if supported, self reset */
+	if (self->sw_reset_supported) {
+		fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+		return fu_cxaudio_device_reset (self, error);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cxaudio_device_set_quirk_kv (FuDevice *device,
+				const gchar *key,
+				const gchar *value,
+				GError **error)
+{
+	FuCxaudioDevice *self = FU_CXAUDIO_DEVICE (device);
+	if (g_strcmp0 (key, "ChipIdBase") == 0) {
+		self->chip_id_base = fu_common_strtoull (value);
+		return TRUE;
+	}
+	if (g_strcmp0 (key, "IsSoftwareResetSupported") == 0) {
+		self->sw_reset_supported = fu_common_strtoull (value);
+		return TRUE;
+	}
+	if (g_strcmp0 (key, "EepromPatchValidAddr") == 0) {
+		self->eeprom_patch_valid_addr = fu_common_strtoull (value);
+		return TRUE;
+	}
+	if (g_strcmp0 (key, "EepromPatch2ValidAddr") == 0) {
+		self->eeprom_patch2_valid_addr = fu_common_strtoull (value);
+		return TRUE;
+	}
+	g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "quirk key not supported");
+	return FALSE;
+}
+
+static void
+fu_cxaudio_device_init (FuCxaudioDevice *self)
+{
+	self->sw_reset_supported = TRUE;
+	fu_device_add_icon (FU_DEVICE (self), "audio-card");
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_install_duration (FU_DEVICE (self), 3); /* seconds */
```

Comment:
Any reason not to use the quirk file with the key `InstallDuration`?

Reply:
The actual install takes <1s, I was adding a couple of seconds for the USB re-enumeration. Even newer devices were 10x slower it'd still be less than 3s. I figured if it's basically a constant, the code would be better to avoid a whole load of string matches at startup.

Reply:
OK makes sense.  Considering this, I wonder if we should take a run over the other plugins to find stuff that is "static" like this and bring it into code too.

Reply:
I don't think it's a performance bottleneck now.